Support mipi dsi on imx95

RuoshanShi opened the pull request:
- Added mipi dsi panel rm692c9 driver support
- Added nxp_mx9_dsi_oled shield support
- Enable the mipi dsi and panel on i.MX95

RuoshanShi committed:
drivers: display: Add RM692C9 driver support

Added rm692c9 driver for i.MX95 display

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi committed:
boards: nxp: enable mipi dsi support on imx95

Enable nxp_mx9_dsi_oled mipi dsi panel

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi committed:
soc: imx95: m7: Config power setting for mipi dsi

- Power on display and camera mix for mipi dsi
- Configure display clock settings
- Add pixel link configuration

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

dleach02 commented:
@EmilioCBen @mmahadevan108 @ngphibang can you guys help review this PR ..

mmahadevan108 commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_m7.dts:
```diff
@@ -148,3 +167,38 @@
+		data-lanes = <4>;
+		pixel-format = <0>;
+	};
+};
```
I would suggest referring to some shields we have for existing displays from raydium and add a similar shield
https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/shields/rk055hdmipi4m
https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/shields/g1120b0mipi

RuoshanShi replied:
The MX9_DSI_OLED shield support has been added

ngphibang commented on drivers/mipi_dsi/Kconfig.mcux:
```diff
@@ -43,3 +43,10 @@ config MIPI_DSI_MCUX_2L_SWAP16
 
 endif # MIPI_DSI_MCUX_2L
+
+config MIPI_DSI_DWC
```
The config also need something like nxp or mcux in the name to distinguish with others

RuoshanShi replied:
Fixed the macro to MIPI_DSI_NXP_DWC in https://github.com/zephyrproject-rtos/zephyr/pull/97587

ngphibang commented on drivers/mipi_dsi/dsi_dwc.c:
```diff
@@ -0,0 +1,202 @@
+		dsi_xfer.dscCmd = msg->cmd;
+		dsi_xfer.txDataType = kDSI_TxDataDcsShortWrOneParam;
+		break;
+
```
no new line

RuoshanShi replied:
Remove the new line in https://github.com/zephyrproject-rtos/zephyr/pull/97587

ngphibang commented on drivers/display/display_rm692c9.c:
```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2025, NXP
```
idem

RuoshanShi replied:
> idem

Fixed

RuoshanShi commented on boards/nxp/imx95_evk/imx95_evk-pinctrl.dtsi:
```diff
@@ -62,6 +62,17 @@
+			input-enable;
+		};
+	};
+
```
@mmahadevan108  No, the LPI2C2 peripheral is responsible for controlling the ADP5585 GPIO expander, enabling the DSICSI_EN_PWDN and DSICSI_RST_SYNC lines by setting them to active high.

RuoshanShi commented on drivers/mipi_dsi/dsi_dwc.c:
```diff
@@ -0,0 +1,202 @@
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/mipi_dsi.h>
+#include <zephyr/logging/log.h>
```
@ngphibang  Fixed them in alphabetic order in https://github.com/zephyrproject-rtos/zephyr/pull/97587

RuoshanShi commented on dts/bindings/mipi-dsi/nxp,mipi-dsi-dwc.yaml:
```diff
@@ -0,0 +1,54 @@
+#
+# Copyright 2025, NXP
```
Fixed the copyright in https://github.com/zephyrproject-rtos/zephyr/pull/97587

RuoshanShi commented on drivers/mipi_dsi/dsi_dwc.c:
```diff
@@ -0,0 +1,202 @@
+	status_t result = DSI_PowerUp(config->base);
+
+	if (result != 0U) {
+		LOG_ERR("DSI PHY init failed.\r\n");
```
Return the result in https://github.com/zephyrproject-rtos/zephyr/pull/97587

RuoshanShi commented on drivers/mipi_dsi/dsi_dwc.c:
```diff
@@ -0,0 +1,202 @@
+/*
```
@mmahadevan108 Yes, the file name has been rename to dsi_nxp_dwc.c in https://github.com/zephyrproject-rtos/zephyr/pull/97587 Thanks for the comment

RuoshanShi commented on drivers/mipi_dsi/dsi_dwc.c:
```diff
@@ -0,0 +1,202 @@
+		(DISPLAY_BLK_CTRL_DISPLAYMIX_PIXEL_LINK_CTRL_PL0_enable(0x1) |
+		 DISPLAY_BLK_CTRL_DISPLAYMIX_PIXEL_LINK_CTRL_PL0_valid(0x1));
+	CAMERA__DSI_OR_CSI_PHY_CSR->COMBO_PHY_MODE_CONTROL = 0x3U;
+#endif
```
Ideally, all related operations should be placed in soc.c. However, since the MIPI DSI HOST register configurations of i.MX93 and i.MX95 are quite similar, but the DPHY configurations differ significantly, platform-specific handling is still required in dsi_nxp_dwc.c.
For details, please refer to: https://github.com/zephyrproject-rtos/zephyr/pull/97587

RuoshanShi committed:
boards: shields: add nxp MX9_DSI_OLED shield

- The shield provides support for MX9_DSI_OLED panel.
- Added shield document.

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi commented on dts/bindings/display/nxp,dpu.yaml:
```diff
@@ -0,0 +1,36 @@
+
+  vbp:
+    type: int
+    description: Vertical back porch in display clock cycles
```
> The display clock cycle properties can be deleted as they can be found in the [dts/bindings/display/panel/panel-timing.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/display/panel/panel-timing.yaml), please reference this for example. https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/shields/rk055hdmipi4ma0/rk055hdmipi4ma0.overlay#L44-L62

Fixed that in boards/shields/nxp_mx9_dsi_oled/nxp_mx9_dsi_oled.overlay

RuoshanShi commented on drivers/display/Kconfig.rm692c9:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2025, NXP
```
> `Copyright 2025 NXP` ?

Fixed the copyright

RuoshanShi commented on drivers/display/display_rm692c9.c:
```diff
@@ -0,0 +1,188 @@
+		}
+		/* Per datasheet, reset low pulse width should be at least 10usec */
+		k_sleep(K_USEC(30));
+		gpio_pin_set_dt(&config->reset_gpio, 1);
```
> It seems you forgot
> 
> `ret = gpio_pin_set_dt(&config->reset_gpio, 1);`
> 
> as you check ret in the next line

thanks , fixed

RuoshanShi commented on drivers/display/display_rm692c9.c:
```diff
@@ -0,0 +1,188 @@
+		.num_of_lanes = DT_INST_PROP_BY_IDX(id, data_lanes, 0),                            \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(id, &rm692c9_init, NULL, NULL, &rm692c9_config_##id, POST_KERNEL,    \
+			      CONFIG_APPLICATION_INIT_PRIORITY, NULL);
```
> Although I see some display panels are initialized with CONFIG_APPLICATION_INIT_PRIORITY, I still wonder if it is too late ? For example, if a subsys / library like libMP (gstreamer like ) wants to call to the driver to set pixelformat, blanking on/off etc., the driver needs to be initialized before it ?
> 
> EDIT: libMP comes later so we will deal with this in another thread, not related to this PR though.
> 
> BTW, the driver just initialize and I don't see any DEVICE_API implemented here ?

1. Used CONFIG_DISPLAY_INIT_PRIORITY for this driver.
2. Added some basic rm692c9 Api

RuoshanShi commented on drivers/display/display_rm692c9.c:
```diff
@@ -0,0 +1,188 @@
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
```
> seems the pm headers are not needed.

Removed the pm headers

dleach02 commented:
@RuoshanShi please address the comments and help move your PR to a merge ready state

RuoshanShi commented on dts/bindings/display/socionext,dpu.yaml:
```diff
@@ -0,0 +1,25 @@
+    enum:
+      - "rgb-565"
+      - "rgb-888"
+      - "argb-8888"
```
> Please use formats defined in `include/zephyr/dt-bindings/mipi_dsi/mipi_dsi.h`

Fixed the code to use the pixel-format in lcd-controller.yml.

ngphibang reviewed (commented):
LGTM but not sure if it is mandatory to add test for the new drivers in `tests/drivers/build_all/display `?

RuoshanShi committed:
tests: drivers: Add RM692C9 device to build test

Add build test for RM692C9 device

Signed-off-by: Ruoshan Shi <ruoshan.shi@nxp.com>

RuoshanShi commented:
@JarmouniA @ngphibang Thanks for the comment. A new commit (f9cfc7fa4bbeddfe139b0ff5992ae0b3e7bda4b3) has been added to support the test. Could you please help review it?

RuoshanShi commented:
@JarmouniA  @ngphibang Any comments or suggestions?

ngphibang reviewed (approved):
LGTM

RuoshanShi commented:
@mmahadevan108 , could you please review this PR?

zejiang0jason commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -98,6 +98,146 @@ static int soc_init(void)
 	}
 #endif
 
+#if defined(CONFIG_MIPI_DSI_NXP_DWC)
```
improvement suggestion in the future: place these in separate small functions and call them in `soc_init`, to improve the readability.

RuoshanShi replied:
> improvement suggestion in the future: place these in separate small functions and call them in `soc_init`, to improve the readability.

Thanks a lot for the suggestion, I will do this in other PR

zejiang0jason commented:
self assigned, this is NXP drivers.
The PR is ok except one small improvement suggestion. @mmahadevan108 's question is answered. So I approve this PR. thanks.
